src/ripple/app/tx/impl/SusPay.cpp

```diff
@@ -391,22 +427,51 @@ SusPayFinish::doApply()
             ctx_.view().info().parentCloseTime.time_since_epoch().count())
         return tecNO_PERMISSION;
 
+    // Check cryptocondition fulfillment
     {
+        auto const id = ctx_.tx.getTransactionID();
+        auto flags = ctx_.app.getHashRouter().getFlags (id);
+
+        auto const cb = ctx_.tx[~sfCondition];
+
+        // It's unlikely that the results of the check will
+        // expire from the hash router, but if it happens,
+        // simply re-run the check.
+        if (cb && ! (flags & (SF_CF_INVALID | SF_CF_VALID)))
+        {
+            auto const fb = ctx_.tx[~sfFulfillment];
+
+            if (!fb)
+                return tecINTERNAL;
+
+            if (checkCondition (*fb, *cb))
+                flags = SF_CF_VALID;
+            else
+                flags = SF_CF_INVALID;
+
+            ctx_.app.getHashRouter().setFlags (id, flags);
+        }
+
+        // If the check failed, then simply return an error
+        // and don't look at anything else.
+        if (flags & SF_CF_INVALID)
+            return tecCRYPTOCONDITION_ERROR;
+
         // Check against condition in the ledger entry:
         auto const cond = (*slep)[~sfCondition];
 
         // If a condition wasn't specified during creation,
         // one shouldn't be included now.
-        if (!cond && ctx_.tx[~sfCondition])
-            return temMALFORMED;
+        if (!cond && cb)
+            return tecCRYPTOCONDITION_ERROR;
 
         // If a condition was specified during creation of
         // the suspended payment, the identical condition
         // must be presented again. We don't check if the
         // fulfillment matches the condition since we did
         // that in preflight.
-        if (cond && (cond != ctx_.tx[~sfCondition]))
-            return temINCORRECT_FULFILLMENT;
+        if (cond && (cond != cb))
+            return tecCRYPTOCONDITION_ERROR;
```

Comment:
Nit: Lines 465 - 474 can all be merged into a single `if (cond != cb)` conditional. No need to separate the cases.